### PR TITLE
Add hold functionaility

### PR DESCRIPTION
**Pull request by 7ritn**
This pull requests will add an option to hold back packages from the queue until the host with the specified IP address (i.e. the one being woken up) responds to a ping. I'm not sure if there is a better way to check for a successful wake-up than with a ping. I believe it's difficult to derive an IP address from a MAC, that's why I chose the need to specify it.

Please respond with any suggestions or feedback. Btw. I believe you can specify in GitHub that Pull Request should be squashed, so I don't have to do the squashing ([GitHub Docs](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests))

### CODE REVIEW

**Comment by mister-benjamin** [comment]
I would probably prefer to not spawn a shell and call `ping`, but do it from within the program... Maybe copying over the `ping4()` function from [busybox](https://git.busybox.net/busybox/tree/networking/ping.c) is a viable solution.

> ...hold back packages from the queue until the host with the specified IP address (i.e. the one being woken up) responds to a ping...

"The one being woken up" needs to be woken up before it can respond to a ping, but it wouldn't be woken up because it doesn't respond to a ping, right?

**Comment by 7ritn** [comment]
> I would probably prefer to not spawn a shell and call ping, but do it from within the program... Maybe copying over the ping4() function from [busybox](https://git.busybox.net/busybox/tree/networking/ping.c) is a viable solution.

Sure I could do that

> "The one being woken up" needs to be woken up before it can respond to a ping, but it wouldn't be woken up because it doesn't respond to a ping, right?

The magic package is always sent **then** if hold is enabled, the package from the queue is held back until the host which hopefully received the magic package woke up and responded to a ping.

**Comment by 7ritn** [comment]
Okay, I've implemented a ping functionality within the project. Ended up modifying a lot from the busybox implementation.

**Comment by mister-benjamin** [comment]
Thanks. I will review in the next days - I am currently short on time.

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,7 +5,9 @@ set(CMAKE_C_STANDARD 99)
 
 add_executable(etherwake-nfqueue
         ether-wake.c
-        nfqueue.c nfqueue.h)
+        nfqueue.c
+        hold.c
+        ping.c)
```
Comment (mister-benjamin) [inline CMakeLists.txt:10]: Maybe one file for the feature is enough, e.g. `hold.c`. But it's not that important.

File: README.md [hunk 2]
```diff
@@ -4,11 +4,11 @@
 ## Wake up computers on netfilter match
 
 **etherwake-nfqueue** is a fork of the **etherwake** Wake-on-LAN client,
-with support to send magic packets only after a queued packet is received
+with support to send magic packets only after a queued send_packet is received
```
Comment (mister-benjamin) [inline README.md]: It seems that your IDE wasn't smart enough and changed all occurrences of the word 'packet' to 'send_packet'.

File: ether-wake.c [hunk 3]
```diff
@@ -23,6 +23,7 @@ static char usage_msg[] =
 "		-b	Send wake-up packet to the broadcast address.\n"
 "		-D	Increase the debug level.\n"
 "		-i ifname	Use interface IFNAME instead of the default 'eth0'.\n"
+"		-h ipaddress	Hold packages back until host with IPADDRESS responds to a ping.\n"
```
Comment (mister-benjamin) [inline ether-wake.c]: Shouldn't this read 'Hold packages back while host responds to a ping' or 'Don't send wake-up packets as long as host with IPADDRESS is up?'

File: ether-wake.c [hunk 4]
```diff
@@ -242,7 +251,11 @@ int main(int argc, char *argv[])
 
 	if (verbose || debug)
 		printf("Acting on packets in NFQUEUE %d\n", opt_nfqueue_num);
-	return nfqueue_receive(opt_nfqueue_num, send_magic_packet);
+
+	ret = nfqueue_receive(opt_nfqueue_num, send_magic_packet);
+
+	cleanup_hold();
```
Comment (mister-benjamin) [inline ether-wake.c]: Maybe this should also be wrapped with `if (hold)`.

File: ether-wake.c [hunk 5]
```diff
@@ -278,6 +291,11 @@ static int send_magic_packet()
 	}
 #endif
 
+	// Wait until host is reachable again
+	if (hold) {
+		hold_for_online();
+	}
```
Comment (mister-benjamin) [inline ether-wake.c]: I think we shouldn't couple the method send_magic_packet() to the higher level concept of hold; I would compose the control flow in the program's main loop - or implement a new function to be used as the callback which does the online check (if option hold is used) and then conditionally call `send_magic_packet()`.

Also, I think that the ping should return right away and not wait for too long, so that the main loop can continue consuming data from the socket.

Btw, for purposes of holding back magic packets (to not unnecessarily send a lot of them), I have some suggestions in the README of the corresponding [OpenWrt package repository](https://github.com/mister-benjamin/etherwake-nfqueue-openwrt#setup-firewall-rules):

> The [iptables] option --match conntrack --ctstate NEW only matches packets of a new connection and --match limit --limit 3/hour --limit-burst 1 limits the amount of packets that are matched. The latter option roughly matches only one packet per 20 minutes. The intention here is to not be too intrusive and avoid sending a lot of magic packets.

File: ether-wake.c (hunk 3 above)
Comment (7ritn) [reply ether-wake.c]: I guess depends on the implementation, currently the program always sends a magic package regardless of ping response. The ping is only used to determine when to release the packet. If the host responds or has responded in the last minute, the packet is released.

File: CMakeLists.txt (hunk 1 above)
Comment (7ritn) [reply CMakeLists.txt:10]: Yeah, I could merge it, but I also think that it's a good idea to separate the higher level wait for online logic from the actual implementation on how to check if the host is online.

File: README.md (hunk 2 above)
Comment (7ritn) [reply README.md]: Thanks, fixed!

File: ether-wake.c (hunk 4 above)
Comment (7ritn) [reply ether-wake.c]: Good idea, fixed!

File: ether-wake.c (hunk 5 above)
Comment (7ritn) [reply ether-wake.c]: > I think we shouldn't couple the method send_magic_packet() to the higher level concept of hold; I would compose the control flow in the program's main loop - or implement a new function to be used as the callback which does the online check (if option hold is used) and then conditionally call send_magic_packet().

Depends on the way one wants to implement this feature. See the other comment. I personally believe it would be better to always send the magic packet, since it would add barely any traffic to the network. And what to do if the host went to sleep, and we still believe it is up, since we got a ping just 10 seconds ago? So we could not implement any assumption and would need to ping the host for every packet, which would delay the release of the packet.
I can implement another callback for hold, which also includes calling the magic packet function.

> Also, I think that the ping should return right away and not wait for too long, so that the main loop can continue consuming data from the socket.

Well, I mean, what would the process do with the consumed packets if it does not know if the host is up or not. Since one instance of etherwake-nfqueue can only handle one destination, the program needs to know if the host is online or not. I'm not sure if asynchronicity would benefit the program?

> Btw, for purposes of holding back magic packets (to not unnecessarily send a lot of them), I have some suggestions in the README of the corresponding [OpenWrt package repository](https://github.com/mister-benjamin/etherwake-nfqueue-openwrt#setup-firewall-rules)

Yeah, that is a good idea to prevent unnecessary magic packages, I personally would increase the limit, but this depends on one's own setup.

File: ether-wake.c (hunk 3 above)
Comment (7ritn) [reply ether-wake.c]: But as I understand it, you would prefer to first ping the host, then if the host does not respond send a magic packet?

**Comment by 7ritn** [comment]
Hi,
do you have any opinion on the comments I made?

**Review by 7ritn — COMMENTED** [review]
Oh my god. I think my remarks were not visible the entire time? Sorry for seeming not to respond.

**Comment by mister-benjamin** [comment]
> Oh my god. I think my remarks were not visible the entire time? Sorry for seeming not to respond.

Yeah, I was wondering... I will have a look and think about it on the weekend.

File: ether-wake.c (hunk 3 above)
Comment (mister-benjamin) [reply ether-wake.c]: Maybe I misunderstood the whole point. Isn’t the “hold” feature a mechanism to hold back magic packets from being sent if the destination host is already online? So I would expect that there’s a check if the host is online and if yes refrain from sending the magic packet and continue normal operation.

Comment (mister-benjamin) [reply ether-wake.c]: If one would only setup the nfqueue client if the host is offline, I think the program would be better composed with the hold mechanism completely decoupled from the wol/nfqueue implementation: wait until the host is offline and only then setup the nfqueue client. After one magic packet has been sent the program can go back pinging the host… A simple flag ‘wakeOnlyOnce’ or similar would be sufficient.

File: ether-wake.c (hunk 5 above)
Comment (mister-benjamin) [reply ether-wake.c]: > I can implement another callback for hold, which also includes calling the magic packet function.

Yeah, that would be nice. Depending on configuration, one or the other could be registered.

> Well, I mean, what would the process do with the consumed packets if it does not know if the host is up or not. Since one instance of etherwake-nfqueue can only handle one destination, the program needs to know if the host is online or not. I'm not sure if asynchronicity would benefit the program?

There’s not necessarily a requirement for concurrency, but as the callback is called from the read loop of the nfqueue socket, I believe the callback shouldn’t block for too long as incoming data on the socket would pile up without being consumed. An easy way around this would be my suggestion in the other comment to set up and tear down the nfqueue client depending on the host’s online state. Otherwise maybe an atomic bool could be updated from a ping thread or only one icmp packet per callback could be sent.

Comment (mister-benjamin) [reply ether-wake.c]: PS: I think not consuming messages from the nfqueue socket also means that the filtered network packets are stuck in the netfilter subsystem until the verdict has been issued, which would also be a strong argument to not block the read loop (for too long). Maybe it would even be better to not send the magic packet in the callback itself, but to only set a flag that it should be send.

Comment (7ritn) [reply ether-wake.c]: Sure, concurrency makes sense, when the packet is only considered an event for etherwake-nfqueue to act on. This is currently the case and there I can see the approach of the packet only activating a flag to send a magic packet.

However, in my use case I would like to hold back packets until I've woken up the host and received a ping. If we were to release the packet before, the host will not receive it and the client has to send a new packet. With my approach, concurrency makes less sense, since all packets in the queue require the host to be up. Thus, it is preferable to hold them in the queue then sending them to an unresponsive host.

File: ether-wake.c (hunk 3 above)
Comment (7ritn) [reply ether-wake.c]: > Maybe I misunderstood the whole point. Isn’t the “hold” feature a mechanism to hold back magic packets from being sent if the destination host is already online? So I would expect that there’s a check if the host is online and if yes refrain from sending the magic packet and continue normal operation.

I personally don't really care if some magic packets are send to a host which is online. I mean the magic packet is 102 Bytes so even sending 1 per second does not impact the network much.

> If one would only setup the nfqueue client if the host is offline, I think the program would be better composed with the hold mechanism completely decoupled from the wol/nfqueue implementation: wait until the host is offline and only then setup the nfqueue client. After one magic packet has been sent the program can go back pinging the host… A simple flag ‘wakeOnlyOnce’ or similar would be sufficient.

I like this idea for reducing the impact of etherwake-nfqueue on network traffic if the host is up, but is unrelated to my idea of holding back client packets until the host is up. I can implement your proposal independent of this PR.

File: ether-wake.c (hunk 5 above)
Comment (mister-benjamin) [reply ether-wake.c]: I think I am beginning to understand now: your whole point is to buffer up any TCP or UDP packets sent to an offline host and release them by sending the nfqueue verdict only after the host is online?

Comment (mister-benjamin) [reply ether-wake.c]: Out of curiosity, what’s exactly your use case? Isn’t that maybe an abuse of the netfilter/nfqueue subsystem with potential problems when it’s taking some time until the host is up and a lot of traffic is coming in? It probably also means that you can’t use any of the iptables rule optimisations like “new connections only” or “rate limiting” and every TCP or UDP packet will pass through etherwake-nfqueue which will send a magic packet and issue a verdict?

**Comment by 7ritn** [comment]
I use this for my home NAS to sleep when not in use and still be always available. I use regular Netfilter Rules such as queueing a packet only when it's initializing a new TCP connection. Works like a charm. I want to hold the packets back so the client gets a reply to its first packet and does not falsely believe that the server is offline.

Sure the option can delay the packet for a few milliseconds, but since this is for home use I don't care that, if the server is online, initial packets take maybe 10ms longer.

-------- Original Message --------

**Comment by mister-benjamin** [comment]
Yeah, I am also using it with my NAS, but everything I use has a retry mechanism.

With TCP, there can of course also be fragmentation, retransmission of packets, or multiple requests from a single service, or simultaneous requests from multiple services. In these cases it wouldn’t help.

If you register a separate callback for the hold mechanism and maybe rephrase the help text to make it more obvious what the option is about (something like “defer delivery of matched packet until host has woken up”), I think we can get this merged. Also, maybe something else than ‘-h’ would be better, as by convention this is used for ‘help’ in most programs.

**Comment by 7ritn** [comment]
Okay perfect. Moved defer mechanism to a new callback and modified help.

File: ether-wake.c [hunk 6]
```diff
@@ -97,10 +98,13 @@ int debug = 0;
 u_char wol_passwd[6];
 int wol_passwd_sz = 0;
 
+static int hold = 0;
+
 static int opt_no_src_addr = 0, opt_broadcast = 0;
 static int opt_nfqueue_num = -1;
 
 static int send_magic_packet();
+static int send_deferred_magic_packet();
```
Comment (mister-benjamin) [inline ether-wake.c]: The name suggests that the function is deferring the magic packet, but that's not what's happening. Could you rename the function to `send_magic_packet_wait_online()` or something?

File: ether-wake.c [hunk 7]
```diff
@@ -23,6 +23,7 @@ static char usage_msg[] =
 "		-b	Send wake-up packet to the broadcast address.\n"
 "		-D	Increase the debug level.\n"
 "		-i ifname	Use interface IFNAME instead of the default 'eth0'.\n"
+"		-d ipaddress	Defer delivery of matched packages until host with IPADDRESS responds to a ping i.e. has woken up.\n"
```
Comment (mister-benjamin) [inline ether-wake.c]: Could you replace the word 'packages' by 'packets' (referring to TCP or UDP packets here). Also, could you add a line break somewhere to make this line approx. as long as the other lines.

**Review by mister-benjamin — COMMENTED** [review]
Can you also please squash the commits? If I do it in Github, I think I would become the author...

File: ether-wake.c [hunk 8]
```diff
@@ -109,17 +113,20 @@ static int get_nfqueue_num(const char *optarg);
...
+	int i, c, ret;
 	struct ether_addr eaddr;
+	int(*send_function)() = &send_magic_packet;
 
 	while ((c = getopt(argc, argv, "bDi:p:q:uvV")) != -1)
 		switch (c) {
 		case 'b': opt_broadcast++;	break;
 		case 'D': debug++;			break;
 		case 'i': ifname = optarg;	break;
+		case 'h': hold++; ip_address = optarg; break;
```
Comment (mister-benjamin) [inline ether-wake.c]: This needs to be changed to 'd' to match the help text.

File: ether-wake.c [hunk 9]
```diff
@@ -109,17 +113,20 @@ static int get_nfqueue_num(const char *optarg);
...
+	char *ip_address;
 	int one = 1;				/* True, for socket options. */
 	int errflag = 0, nfqueue_errflag = 0, verbose = 0, do_version = 0;
 	int perm_failure = 0;
-	int i, c;
+	int i, c, ret;
 	struct ether_addr eaddr;
+	int(*send_function)() = &send_magic_packet;
 
 	while ((c = getopt(argc, argv, "bDi:p:q:uvV")) != -1)
```
Comment (mister-benjamin) [inline ether-wake.c]: The getopt string needs to be updated to include the new option. Also, I didn't see any validation for the ip address argument, but never mind... Only a potential improvement.

**Comment by 7ritn** [comment]
> Can you also please squash the commits? If I do it in Github, I think I would become the author...

I believe authorship should remain the same only you will be listed as the person who committed, which is fine.

**Comment by mister-benjamin** [comment]
Thanks! Merged.
